File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/javascript/exercises/practice/affine-cipher/affine-cipher.js
const gcd = (a, b) => {
  while (b !== 0) {
    [a, b] = [b, a % b];
  }
  return a;
};

const areCoprime = (a, m) => gcd(a, m) === 1;

const modInverse = (a, m) => {
  a = ((a % m) + m) % m; // Ensure a is positive
  for (let x = 1; x < m; x++) {
    if ((a * x) % m === 1) {
      return x;
    }
  }
  throw new Error('Modular inverse does not exist');
};

const processText = (text) => text.toLowerCase().replace(/[^a-z0-9]/g, '');

const groupText = (text) => {
  const chunks = [];
  for (let i = 0; i < text.length; i += 5) {
    chunks.push(text.slice(i, i + 5));
  }
  return chunks.join(' ');
};

export const encode = (phrase, key) => {
  const { a, b } = key;
  if (!areCoprime(a, 26)) {
    throw new Error('a and m must be coprime.');
  }

  const processed = processText(phrase);
  let encrypted = '';
  for (const char of processed) {
    if (/\d/.test(char)) {
      encrypted += char;
    } else {
      const i = char.charCodeAt(0) - 97;
      const e = (a * i + b) % 26;
      encrypted += String.fromCharCode(e + 97);
    }
  }

  return groupText(encrypted);
};

export const decode = (phrase, key) => {
  const { a, b } = key;
  if (!areCoprime(a, 26)) {
    throw new Error('a and m must be coprime.');
  }

  const processed = processText(phrase);
  const aInv = modInverse(a, 26);
  let decrypted = '';

  for (const char of processed) {
    if (/\d/.test(char)) {
      decrypted += char;
    } else {
      const y = char.charCodeAt(0) - 97;
      // Calculate (aInv * (y - b)) mod 26, ensuring no negative numbers
      const bMod = ((b % 26) + 26) % 26;
      const d = (aInv * (y - bMod + 26)) % 26;
      decrypted += String.fromCharCode(d + 97);
    }
  }

  return decrypted;
};

File: src/2025-02-24-deepclaude-rust/python/exercises/practice/affine-cipher/affine_cipher.py
def encode(plain_text, a, b):
    import math
    if math.gcd(a, 26) != 1:
        raise ValueError("a and m must be coprime.")
    
    # Process the text: lowercase and filter out non-alphanumeric
    processed = [char for char in plain_text.lower() if char.isalpha() or char.isdigit()]
    
    # Encrypt each character
    encrypted = []
    for char in processed:
        if char.isalpha():
            # Convert letter to number (0-25), apply formula, convert back to letter
            x = ord(char) - ord('a')
            e = (a * x + b) % 26
            encrypted.append(chr(e + ord('a')))
        else:
            # Keep digits as is
            encrypted.append(char)
    
    # Group into chunks of 5
    result = ""
    for i in range(0, len(encrypted), 5):
        if i > 0:
            result += " "
        result += "".join(encrypted[i:i+5])
    
    return result


def decode(ciphered_text, a, b):
    import math
    if math.gcd(a, 26) != 1:
        raise ValueError("a and m must be coprime.")
    
    # Find modular multiplicative inverse of a mod 26
    a_inv = 0
    for i in range(1, 26):
        if (a * i) % 26 == 1:
            a_inv = i
            break
    
    # Remove spaces from ciphered text
    processed = ciphered_text.replace(" ", "")
    
    # Decrypt each character
    decrypted = []
    for char in processed:
        if char.isalpha():
            # Convert letter to number, apply inverse formula, convert back to letter
            y = ord(char) - ord('a')
            d = (a_inv * (y - b)) % 26
            decrypted.append(chr(d + ord('a')))
        else:
            # Keep digits as is
            decrypted.append(char)
    
    return "".join(decrypted)
